nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/ASURLClassLoader.java

```diff
@@ -118,6 +118,9 @@ public class ASURLClassLoader extends GlassfishUrlClassLoader
 
     private final ArrayList<ClassFileTransformer> transformers = new ArrayList<>(1);
 
+    /** Bytecode transformation flag */
+    private static final ThreadLocal<Boolean> BYTECODE_TRANSFORMATION = withInitial(() -> false);
```

Comment:
I think that thread locals should be removed at some point, at least when the classloader is closed. Then they shouldn't be static as classes are defined by the classloader. SonarQube told me :-)
Btw why do we use ThreadLocal here? findClass is called by ClassLoader.loadClass in a synchronized block (synced via getClassLoadingLock).
It seems to me that now it causes just a side effect with multiple threads ... but I can't imagine all situations, because the getClassLoadingLock logic is a bit "magic" for me.

Reply:
@dmatej, yo're absolute right! Because this CL is not registered as a parallel-capable, they obtain itself as lock in the `getClassLoadingLock`. And I think it's enough a simple non-volatile boolean flag.